spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala

```diff
@@ -556,7 +556,24 @@ object Engine {
   }
 
   private def setMklDnnEnvironments(): Unit = {
-    val default = Math.ceil(Runtime.getRuntime.availableProcessors().toFloat / 2).toInt
+    import com.intel.analytics.bigdl.mkl.hardware.CpuInfo
+    val affinityCores = Affinity.getAffinity
+    val physicalCoreNum = CpuInfo.getPhysicalProcessorCount
+    val affinityCoreNum = affinityCores.length
+
+    // 1. this library in docker/cgroup env, which sets cpu affinity fist. so we can't use
+    //    resources exceeding limits.
+    // 2. this library is in a hyper thread envs, which enables hyper threading of cpu. so
+    //    we should set the mkl num threads to physical core number for performance
+
+    val default = if (affinityCores.min > 0 && affinityCores.max >= physicalCoreNumber) {
```

Comment:
When it happens `affinityCores.max >= physicalCoreNumber`

Reply:
hyper threading is enabled.

Reply:
ok, why not use physicalCoreNumber to bind affinity in this case.

Reply:
this case can't use. for example there're 8 logical cores and 4 physical cores. If the affinity settings is 2-6 (starting from 0), which means this environment only can use core 2, 3, 4. So I can't set the `physicalCoreNumber`.

Reply:
ok， I see